### PR TITLE
Multithreaded Benchmarks

### LINKED ISSUE #1005: Multithreaded Benchmarks
The benchmark runner is a good tool to evaluate our performance. Currently, it only executes one query at a time (which might be executed with multiple threads). @lanice built some [cool graphs](https://github.com/hyrise/hyrise/pull/994#issuecomment-408680431) for multi-threaded multi-client benchmarks.

It should be possible to run this as easily as the current benchmarks are and get a similar graph. Because this only makes sense with a plot, I think we should not embed this in the runner, but have a wrapper script for the runner.

Interesting command line options:
- number of clients (i.e., queries running in parallel)
- number of queries / max execution time
- thread stepping (for a quick evaluation, it should be enough to evaluate 1, 2, 3, 4, 5, 10, 15, ... threads)

Before we start to work on this, let's discuss it.

### PR DESCRIPTION
**Pull request by lanice**
Closes #1005.

This adds the parameters `--cores X` and `--clients X` to the benchmarkRunner, with `cores` specifying the number of cores that will be used IF `--scheduler` is active, and `clients` setting the number of simulated clients that schedule queries in parallel.

This also adds a python script, `scripts/benchmark_multithreaded.py`, which allows for benchmarking over multiple numbers of cores and putting the results together into a plot. Currently for the TPC-H benchmark only.

The python script takes the following parameters:
- `EXECUTABLE`: mandatory, the path to the `hyriseBenchmarkTPCH` executable
- `--runs X` OR `--runs-per-core X`: one of the two has to be set. `--runs X` means, no matter the number of cores, we always pass `--runs X` to the Hyrise benchmark. `--runs-per-core X` means we pass `--runs Y` to the Hyrise benchmark, with Y being X times the number of cores of the current benchmark run
- `--max-cores X` OR `--cores X Y Z`: optional. If neither is set, the benchmark will be executed for a range of cores from 1 to the maximum available number of cores. `--cores X Y Z` specifies which core counts exactly should be benchmarked
- `--queries X Y Z`: optional. If not set, all possible TPC-H queries are benchmarked, otherwise the specified
- `--result-dir` and `result-name`: optional, defaults are `results/` and current datetime
- `--verbose`: print debug output
- all other given parameters (like `scale`, `chunk_size`, etc) will be passed on to the Hyrise benchmark

Here an example execution:
```
python3 ./scripts/benchmark_multithreaded.py ./build-release/hyriseBenchmarkTPCH --runs 1000 -v
```
The final plot could look like this:
plots1

And another example, this time a bit more specified:
```
python3 ./scripts/benchmark_multithreaded.py ./build-release/hyriseBenchmarkTPCH --runs 1000 -v --scale 1 --cores 10 20 30 --queries 1 3 6 12
```
Possible output plot:
plots2

### CODE REVIEW

File: scripts/benchmark_multithreaded.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-s', '--scale', action='store', type=float, metavar='S', default=0.1, help='TPC-H scale factor (default: 0.1)')
+    parser.add_argument('-v', '--verbose', action='store_true', help='Print log messages')
+    parser.add_argument('-q', '--queries', action='store', type=int, metavar='Q', nargs='+', help='Specify the TPC-H queries that will be benchmarked')
+    parser.add_argument('--chunk-size', action='store', type=int, metavar='S', help='Specify maximum chunk size (default: Maximum available)')
+    parser.add_argument('--result-dir', action='store', type=str, metavar='DIR', default='results', help='Directory where the result folder will be stored (default: \'results/\')')
+    parser.add_argument('--result-name', action='store', type=str, metavar='NAME', help='Directory where the actual results will be stored (default: current datetime)')
+    parser.add_argument('--build-dir', action='store', type=str, metavar='DIR', required=True, help='Directory that contains the hyriseBenchmarkTPCH executable')
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]: Why do you take the build directory instead of the binary? How about this syntax, which allows us to pass any type of arguments to the binary without explicitly listing them in this script:

`benchmark_multithreaded.py [--script-options=foo] ./build-release/hyriseBenchmarkTPCH [--hyrise_options]`

File: scripts/benchmark_multithreaded.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+def run_benchmarks(args, core_counts, executable, result_dir):
+    benchmark_run = 0
+    for core_count in core_counts:
+        benchmark_run += 1
+        verbose_print(args.verbose, 'Starting benchmark run ' + str(benchmark_run) + ' of ' + str(len(core_counts)) + ' ...')
+
+        if core_count == 0:
+            use_scheduler = 'false'
+            number_of_runs = args.runs_per_core if args.runs_per_core else args.fixed_runs
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]: Thinking about this again, I still think this is more complicated than it has to be. What if we allow `--runs=0` on the Hyrise side which means "run until time is up". Wouldn't that solve the problem here?

File: src/benchmarklib/benchmark_runner.cpp [hunk 3]
```diff
@@ -103,21 +104,44 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
   std::random_device random_device;
   std::mt19937 random_generator(random_device());
 
-  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
-  while (state.keep_running()) {
-    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
+  if (_config.parallel_execution) {
```
Comment (mrks) [inline src/benchmarklib/benchmark_runner.cpp]: If `--parallel_clients` took an integer, we wouldn't need this `if`, correct?

File: src/benchmarklib/benchmark_utils.cpp [hunk 4]
```diff
@@ -140,7 +143,21 @@ BenchmarkConfig CLIConfigParser::parse_basic_options_json_config(const nlohmann:
   out << "- MVCC is " << (enable_mvcc ? "enabled" : "disabled") << std::endl;
 
   const auto enable_scheduler = json_config.value("scheduler", default_config.enable_scheduler);
-  out << "- Running in " + std::string(enable_scheduler ? "multi" : "single") + "-threaded mode" << std::endl;
+  const auto available_cores = json_config.value("cores", default_config.available_cores);
```
Comment (mrks) [inline src/benchmarklib/benchmark_utils.cpp]: This should verify that the scheduler is enabled

File: scripts/benchmark_multithreaded.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+BENCHMARK_EXECUTABLE = 'hyriseBenchmarkTPCH'
+DEFAULT_TPCH_QUERIES = [query for query in range(1, 23) if query != 15] # Exclude query 15 which is not supported in our multithreaded benchmarks
+MAX_CORE_COUNT = multiprocessing.cpu_count()
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-s', '--scale', action='store', type=float, metavar='S', default=0.1, help='TPC-H scale factor (default: 0.1)')
+    parser.add_argument('-v', '--verbose', action='store_true', help='Print log messages')
+    parser.add_argument('-q', '--queries', action='store', type=int, metavar='Q', nargs='+', help='Specify the TPC-H queries that will be benchmarked')
```
[the remark this answers is not in the document]
Comment (lanice) [reply scripts/benchmark_multithreaded.py]: To your first question: No, `--queries 1 3 5` is the way python argparse handles "list" parameters, comma-separated doesn't work.

[the remark this answers is not in the document]
Comment (lanice) [reply scripts/benchmark_multithreaded.py]: To your second question, that is unfortunately a `cxxopt` problem, which basically cannot handle "list" inputs like that. You'd have to use the `--queries` parameter for each query you want to add:
```
./build/hyriseBenchmarkTPCH --queries 1 --queries 2 -s 0.001
```
That's why I added the shortcut `-q` for `--queries` there, because it makes life much better when you instead can just do:
```
./build/hyriseBenchmarkTPCH -q 1 -q 2 -s 0.001
```

[the remark this answers is not in the document]
Comment (mrks) [reply scripts/benchmark_multithreaded.py]: Maybe accept a string in both cases and split it manually?

**Comment by lanice** [comment]
@mrks As far as the cpp code goes, I addressed the review comments and what we talked about in person, except the one with the cxxopts parameter to specify queries.
I will adjust the python script as well, but if you want, you can already have a look at the hyrise code.

File: src/benchmarklib/benchmark_runner.cpp [hunk 6]
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
...
+  auto currently_running_query_sets = std::atomic_uint{0};
+  auto finished_query_set_runs = std::atomic_uint{0};
+  auto finished_queries_total = std::atomic_uint{0};
+  // For the iteration durations vector, we need a mutex
+  auto result_mutex = std::mutex{};
+
+  auto tasks = std::vector<std::shared_ptr<AbstractTask>>{};
+  auto state = BenchmarkState{_config.max_num_query_runs, _config.max_duration};
+
+  while (state.keep_running(finished_query_set_runs.load(std::memory_order_relaxed))) {
```
Comment (mrks) [inline src/benchmarklib/benchmark_runner.cpp]: I don't understand this line

File: src/benchmarklib/benchmark_runner.cpp [hunk 7]
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
...
+
+      for (const auto& named_query : mutable_named_queries) {
+        // The on_query_done callback will be appended to the last Task of the query,
+        // to measure its duration as well as signal that the query was finished
+        const auto query_run_begin = std::chrono::steady_clock::now();
+        auto on_query_done = [query_run_begin, named_query, number_of_queries_per_set, &currently_running_query_sets,
+                              &finished_query_set_runs, &finished_queries_total, &result_mutex, this]() {
+          const auto duration = std::chrono::steady_clock::now() - query_run_begin;
+          finished_queries_total++;
+          if (finished_queries_total.load(std::memory_order_relaxed) % number_of_queries_per_set == 0) {
```
Comment (mrks) [inline src/benchmarklib/benchmark_runner.cpp]: Why not reuse the return value of `++`?

File: src/benchmarklib/benchmark_runner.cpp [hunk 8]
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
...
-
-      // Execute the query, we don't care about the results
-      _execute_query(named_query);
-
-      const auto query_benchmark_end = std::chrono::steady_clock::now();
-
-      auto& query_benchmark_result = _query_results_by_query_name[named_query.first];
-      query_benchmark_result.duration += query_benchmark_end - query_benchmark_begin;
-      query_benchmark_result.num_iterations++;
+  const auto number_of_queries_per_set = mutable_named_queries.size();
```
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: With `--mode PermutedQuerySets` we are running on sets of queries, not individual queries. So "set" is one (shuffled) batch of the queries we want to run.

File: src/benchmarklib/benchmark_runner.cpp [hunk 9]
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
...
+    // Wait for the rest of the tasks that didn't make it in time - they will not count toward the results
+    // TODO(leander/anyone): To be replaced with something like CurrentScheduler::abort(),
+    // that properly removes all remaining tasks from all queues, without having to wait for them
+    CurrentScheduler::wait_for_tasks(tasks);
+    DebugAssert(currently_running_queries.load(std::memory_order_relaxed) == 0,
+                "All query runs must be finished at this point");
+  }
+}
+
+std::vector<std::shared_ptr<AbstractTask>> BenchmarkRunner::_schedule_or_execute_query(
```
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: The only reason for that is that for Queries that require execution (like TPC-H 15), we can't call get_tasks() on the pipeline to schedule each task. So with that, these queries don't break when we don't use the scheduler.

[the remark this answers is not in the document]
Comment (mrks) [reply src/benchmarklib/benchmark_runner.cpp]: Ok, maybe leave a comment for that.

File: src/benchmarklib/benchmark_runner.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/benchmarklib/benchmark_runner.cpp]: Got it. I think I find "sets" a bit misleading, as there is only one set of queries that is executed in a random order, isn't there?

File: src/benchmarklib/benchmark_runner.cpp (hunk 6 above)
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: Which part?
1) `state.keep_running()` now takes a parameter of how many query (sets) are already finished (will document that in the BenchmarkState declaration)
2) `.load()` is to get the actual value from std::atomic
3) `std::memory_order_relaxed` tells `load` that there are no synchronization or ordering constraints imposed on other reads or writes, only this operation's atomicity is guaranteed. Which, according to [this](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering), should be fine here. However after thinking more about it now, I will reevaluate the memory_ordering uses in the other cases as well.

Comment (mrks) [reply src/benchmarklib/benchmark_runner.cpp]: The first. Even with the documentation, I think that the signature of that method is everything but easy to understand. Why do I have to pass a number into the state to find out if the benchmark should continue? Can't we do the check outside of `keep_running`, i.e.

`while(state.keep_running() && ...`?

File: src/benchmarklib/benchmark_runner.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: Ah I see what you mean. So basically rename the mode PermutedQuerySets -> PermutedQuerySet ?

And actually, `number_of_queries_per_set` might as well be just `number_of_queries`, which is what that is.

File: src/benchmarklib/benchmark_runner.cpp (hunk 6 above)
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: Yes we can. This would leave the BenchmarkState as a timer basically.

File: src/benchmarklib/benchmark_runner.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/benchmarklib/benchmark_runner.cpp]: Personally, I would find that easier to understand

File: src/benchmarklib/benchmark_runner.cpp (hunk 7 above)
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: Yeah that... makes sense...

**Comment by Bouncner** [comment]
Any changes to the examples you provided in the PR description?

**Comment by lanice** [comment]
I haven't updated the python script yet, so the examples won't work right now.

File: scripts/benchmark_multithreaded.py (hunk 2 above)
Comment (lanice) [reply scripts/benchmark_multithreaded.py]: Not sure what problem you mean here. If I specify `--runs=0` (and if that's allowed on the Hyrise side), then that will be used and passed to Hyrise. But if I specify `--runs-per-core=X`, then that instead will be used to determine the number of max query runs for this benchmark run.

**Comment by Bensk1** [comment]
What's the status on that?

**Comment by lanice** [comment]
> What's the status on that?

The PR is done, I'm just waiting for a review.

File: scripts/benchmark_multithreaded.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
+# To prevent _tkinter.TclError: https://stackoverflow.com/a/37605654
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]: add shebang, set `chmod +x` for this file

File: scripts/benchmark_multithreaded.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+import matplotlib.pyplot as plt
+
+import os
+import argparse
+import multiprocessing
+import subprocess
+import json
+
+from datetime import datetime
+
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]:
```
Markus.Dreseler@nemea:~/hyrise/build-release$ python3 ../scripts/benchmark_multithreaded.py --runs 1000 -q 1 2 ./hyriseBenchmarkTPCH
usage: benchmark_multithreaded.py [-h] [-v] [-q Q [Q ...]] [--result-dir DIR]
                                  [--result-name NAME] [-c C [C ...] |
                                  --max-cores C] (-r N | --runs-per-core N)
                                  EXECUTABLE
benchmark_multithreaded.py: error: argument -q/--queries: invalid int value: './hyriseBenchmarkTPCH'
```

Can we do anything about this?

File: scripts/benchmark_multithreaded.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+DEFAULT_TPCH_QUERIES = [query for query in range(1, 23) if query != 15] # Exclude query 15 which is not supported in our multithreaded benchmarks
+
+def get_parser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-v', '--verbose', action='store_true', help='Print log messages')
+    parser.add_argument('-q', '--queries', action='store', type=int, metavar='Q', nargs='+', help='Specify the TPC-H queries that will be benchmarked')
+    parser.add_argument('--result-dir', action='store', type=str, metavar='DIR', default='results', help='Directory where the result folder will be stored (default: \'results/\')')
+    parser.add_argument('--result-name', action='store', type=str, metavar='NAME', help='Directory where the actual results will be stored (default: current datetime)')
+    parser.add_argument('executable', action='store', type=str, metavar='EXECUTABLE', help='hyriseBenchmarkTPCH executable')
+    parser.add_argument('--clients', action='store', type=int, metavar='C', default=10, help='The number of clients that schedule queries in parallel')
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]: This doesn't show up in the help

File: scripts/benchmark_multithreaded.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+def get_subplot_row_and_column_count(num_plots):
+    while not is_square(num_plots):
+        num_plots += 1
+    return num_plots ** 0.5
+
+def run_benchmarks(args, hyrise_args, core_counts, result_dir):
+    benchmark_run = 0
+    for core_count in core_counts:
+        benchmark_run += 1
+        verbose_print(args.verbose, 'Starting benchmark run ' + str(benchmark_run) + ' of ' + str(len(core_counts)) + ' ...')
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]: I would leave out "run" here. In the context of benchmarks, I understand it as if one benchmark was executed multiple times.

File: scripts/benchmark_multithreaded.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+        assert core_count <= MAX_CORE_COUNT, 'Can\'t use more cores than we have available'
+
+    run_benchmarks(args, hyrise_args, core_counts, result_dir)
+
+    verbose_print(args.verbose, 'Benchmarks complete!')
+    verbose_print(args.verbose, 'Saved JSON results to: ' + result_dir)
+
+    return result_dir
+
+def plot(args, result_dir):
```
Comment (mrks) [inline scripts/benchmark_multithreaded.py]: Can we get the NUMA boundaries in?

Comment (mrks) [inline scripts/benchmark_multithreaded.py]: I would add markers to show which part of the line were actually measured

File: src/benchmarklib/benchmark_utils.hpp [hunk 15]
```diff
@@ -113,6 +110,8 @@ struct BenchmarkConfig {
   const UseMvcc use_mvcc = UseMvcc::No;
   const std::optional<std::string> output_file_path = std::nullopt;
   const bool enable_scheduler = false;
+  const uint cores = 0;
+  const uint clients = 1;
```
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmarklib/benchmark_utils.hpp]: The reason behind this is that, if you do specify `--scheduler` but not `--clients`, you basically run the "normal" benchmarks with active scheduler, as we had before when we just added `--scheduler` to the benchmark, executing one benchmark after another, but with active scheduler (and now customizable cores). Also, if the default is != 1, then we'd have to check during setup if the scheduler is active, and set clients back to 1 for the "normal" benchmarks.
Which is fine by me, let me know what you prefer.

File: scripts/benchmark_multithreaded.py (hunk 11 above)
Comment (lanice) [reply scripts/benchmark_multithreaded.py]: We could make the executable a named argument `-e, --executable` instead of a positional argument?

File: scripts/benchmark_multithreaded.py (hunk 12 above)
Comment (lanice) [reply scripts/benchmark_multithreaded.py]: The help for `--clients`? For me it does:

```
lanice:opossumdb (leander/1005_multithreaded_benchmarks) $ ./scripts/benchmark_multithreaded.py --help
usage: benchmark_multithreaded.py [-h] [-v] [-q Q [Q ...]] [--result-dir DIR]
                                  [--result-name NAME] [--clients C]
                                  [-c C [C ...] | --max-cores C]
                                  (-r N | --runs-per-core N)
                                  EXECUTABLE

positional arguments:
  EXECUTABLE            hyriseBenchmarkTPCH executable

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         Print log messages
  -q Q [Q ...], --queries Q [Q ...]
                        Specify the TPC-H queries that will be benchmarked
  --result-dir DIR      Directory where the result folder will be stored
                        (default: 'results/')
  --result-name NAME    Directory where the actual results will be stored
                        (default: current datetime)
  --clients C           The number of clients that schedule queries in
                        parallel
  -c C [C ...], --cores C [C ...]
                        List of cores to be used for the benchmarks
  --max-cores C         Number of cores up to which the benchmarks will be
                        executed
  -r N, --runs N        Fixed number of (max) runs each query is executed
  --runs-per-core N     Number of (max) runs per core each query is executed
```

File: src/benchmarklib/benchmark_runner.cpp [hunk 16]
```diff
@@ -103,40 +104,106 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
...
-      query_benchmark_result.num_iterations++;
+  const auto number_of_queries = mutable_named_queries.size();
+
+  // The atomic uints are modified by other threads when finishing a query set, to keep track of when we can
+  // let a simulated client schedule the next set, as well as the total number of finished query sets so far
+  auto currently_running_query_sets = std::atomic_uint{0};
+  auto finished_query_set_runs = std::atomic_uint{0};
+  auto finished_queries_total = std::atomic_uint{0};
+  // For the iteration durations vector, we need a mutex
+  auto result_mutex = std::mutex{};
```
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: You mean making the `num_iterations` and `duration` atomic and `iteration_durations` into a concurrent vector? But we wouldn't need to pre-allocate elements then, right?
```
// benchmark_utils.hpp

struct QueryBenchmarkResult {
  size_t num_iterations = 0;
  Duration duration = Duration{};
  std::vector<Duration> iteration_durations;
};
```

File: scripts/benchmark_multithreaded.py (hunk 14 above)
Comment (lanice) [reply scripts/benchmark_multithreaded.py]: Good point :+1:

Comment (lanice) [reply scripts/benchmark_multithreaded.py]: Yep, I added the number of cores per NUMA node to the result JSON output, and with that the python script can print the boundaries (see updated image below).

**Comment by lanice** [comment]
Here a sample output plot after the last changes:
result_plots

File: scripts/benchmark_multithreaded.py (hunk 12 above)
Comment (mrks) [reply scripts/benchmark_multithreaded.py]: Maybe I am blind. Or I was on the wrong branch. Or... some other lame excuse.

File: scripts/benchmark_multithreaded.py (hunk 11 above)
Comment (mrks) [reply scripts/benchmark_multithreaded.py]: hmpf. That would mean that if we want to pass parameters to the executable, we would have to enclose it in quotes, right?

And what if, instead of using Python conventions, we use the same query parameter format that Hyrise uses?

Comment (mrks) [reply scripts/benchmark_multithreaded.py]: hmpf. That would mean that if we want to pass parameters to the executable, we would have to enclose it in quotes, right?

And what if, instead of using Python conventions, we use the same query parameter format that Hyrise uses?

File: src/benchmarklib/benchmark_runner.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (mrks) [reply src/benchmarklib/benchmark_runner.cpp]: > But we wouldn't need to pre-allocate elements then

We wouldn't have to, but the default growth algorithm means that we get a lot of fragmentation in the beginning if we don't

File: scripts/benchmark_multithreaded.py (hunk 11 above)
Comment (mrks) [reply scripts/benchmark_multithreaded.py]: Sounds good.

File: src/benchmarklib/benchmark_runner.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmarklib/benchmark_runner.cpp]: Okay, so for some reason I can not get it to work so far, might need some help. When I change just one member (`num_iterations`) of the struct to atomic:
```
struct QueryBenchmarkResult {
  std::atomic<size_t> num_iterations = 0;
  Duration duration = Duration{};
  std::vector<Duration> iteration_durations;
};
```
I get the following error when emplacing/inserting into the result map (`std::unordered_map<std::string, QueryBenchmarkResult>`):
```
...
/home/lanice/develop/epic/opossumdb/src/benchmarklib/benchmark_runner.cpp:105:70:   required from here
/usr/include/c++/7/ext/new_allocator.h:136:4: error: no matching function for call to ‘std::pair<const std::__cxx11::basic_string<char>, opossum::QueryBenchmarkResult>::pair(const std::__cxx11::basic_string<char>&, opossum::QueryBenchmarkResult)’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/utility:70:0,
                 from /usr/include/c++/7/algorithm:60,
                 from /home/lanice/develop/epic/opossumdb/third_party/json/json.hpp:32,
                 from /home/lanice/develop/epic/opossumdb/src/benchmarklib/benchmark_runner.cpp:1:
/usr/include/c++/7/bits/stl_pair.h:431:9: note: candidate: template<class ... _Args1, long unsigned int ..._Indexes1, class ... _Args2, long unsigned int ..._Indexes2> std::pair<_T1, _T2>::pair(std::tuple<_Args1 ...>&, std::tuple<_Args2 ...>&, std::_Index_tuple<_Indexes1 ...>, std::_Index_tuple<_Indexes2 ...>)
         pair(tuple<_Args1...>&, tuple<_Args2...>&,
         ^~~~
/usr/include/c++/7/bits/stl_pair.h:431:9: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/7/bits/c++allocator.h:33:0,
                 from /usr/include/c++/7/bits/allocator.h:46,
                 from /usr/include/c++/7/string:41,
                 from /usr/include/c++/7/stdexcept:39,
                 from /usr/include/c++/7/array:39,
                 from /home/lanice/develop/epic/opossumdb/third_party/json/json.hpp:33,
                 from /home/lanice/develop/epic/opossumdb/src/benchmarklib/benchmark_runner.cpp:1:
/usr/include/c++/7/ext/new_allocator.h:136:4: note:   types ‘std::tuple<_Tps ...>’ and ‘const std::__cxx11::basic_string<char>’ have incompatible cv-qualifiers
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
...
```
Caused when constructing the pair in the `emplace` calls in:
```
// Init results
for (const auto& named_query : _queries) {
  const auto& name = named_query.first;
  _query_results_by_query_name.emplace(name, QueryBenchmarkResult{});
}
```
And:
```
auto result = QueryBenchmarkResult{};
result.num_iterations = finished_query_runs;
result.duration = state.benchmark_duration;
result.iteration_durations = iteration_durations;
_query_results_by_query_name.emplace(name, result);
```

[the remark this answers is not in the document]
Comment (mrks) [reply src/benchmarklib/benchmark_runner.cpp]: You can't copy an atomic, so you need to find a way (references/moving) where you don't need to.

File: scripts/benchmark_multithreaded.py [hunk 17]
```diff
@@ -28,7 +28,7 @@
 # python3 ./scripts/benchmark_multithreaded.py -e ./build-release/hyriseBenchmarkTPCH --runs 1000 -v --scale 1 --cores 10 20 30 --queries 1,3,6,12 --clients 20
 
 
-MAX_CORE_COUNT = multiprocessing.cpu_count()
```
Comment (lanice) [inline scripts/benchmark_multithreaded.py]: Thanks for your changes! I agree with everything except this one. The Topology already handles this if we are on a non-NUMA system. On a NUMA-system it uses all available CPUs. I don't think we should restrict this in two different spots, if it has to go through one of them anyway. Also this would easily get overridden when specifying --cores or --max-cores, while still possibly being changed by the Topology.

File: scripts/benchmark_multithreaded.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+# You can also specify the cores to be benchmarked, otherwise a default range will be used.
+#
+# Example usage 1:
+# python3 ./scripts/benchmark_multithreaded.py -e ./build-release/hyriseBenchmarkTPCH --runs 1000 -v
+#
+# Example usage 2:
+# python3 ./scripts/benchmark_multithreaded.py -e ./build-release/hyriseBenchmarkTPCH --runs 1000 -v --scale 1 --cores 10 20 30 --queries 1,3,6,12 --clients 20
+
+
+MAX_CORE_COUNT = max(1, multiprocessing.cpu_count() - 1)  # Keep one core empty, same as in topology.cpp
```
Comment (lanice) [inline scripts/benchmark_multithreaded.py]: Thanks for your changes! I agree with everything except this one. The Topology already handles this if we are on a non-NUMA system. On a NUMA-system it uses all available CPUs. I don't think we should restrict this in two different spots, if it has to go through one of them anyway. Also this would easily get overridden when specifying --cores or --max-cores, while still possibly being changed by the Topology.

Comment (lanice) [reply scripts/benchmark_multithreaded.py]: Okay, that makes sense. After thinking about it, I honestly think we should remove this limitation completely. If I want to benchmark something on my laptop, with 4 cores, and go `hyriseBenchmarkTPCH --cores 4 ...`, then I want to benchmark it with 4 cores, and not with 3. If I want to leave one core out because I want to work on something else while benchmarking, then I can specify that manually. We don't restrict CPUs on a NUMA system either, if we say use 80 cores, we do it.
Basically I don't want the Benchmark to "not allow" me to fully utilize my system for benchmarking if I want to... right?

Comment (lanice) [reply scripts/benchmark_multithreaded.py]: Which btw, funnily, I did the whole time before. Since I installed libnuma some time ago to test something, CMake thinks I'm on a NUMA system, so I never even got that limitation when benchmarking on my laptop, because I always had a "NUMA" Topology with only 1 node and 4 cores. Worked perfectly, no freeze. Of course I didn't do anything else during that benchmark, but that was my intent, because I knew that all 4 cores would be used.
That's why I forgot about that limitation, because it didn't come up when I tested it ^^

Comment (mrks) [reply scripts/benchmark_multithreaded.py]: Agreed. Please get rid of it.

File: src/benchmarklib/benchmark_runner.cpp [hunk 19]
```diff
@@ -103,52 +111,157 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
...
+        const auto query_run_begin = std::chrono::steady_clock::now();
+        auto on_query_done = [query_run_begin, named_query, number_of_queries, &currently_running_clients,
+                              &finished_query_set_runs, &finished_queries_total, &state, this]() {
+
+          if (finished_queries_total++ % number_of_queries == 0) {
+            currently_running_clients--;
+            finished_query_set_runs++;
+          }
+
+          if (!state.is_done()) { // To prevent queries to add their results after the time is up
```
Comment (lanice) [inline src/benchmarklib/benchmark_runner.cpp]: I added this, because after the time is up we still "wait" for the currently running clients to finish, but of course we don't want them to add their results anymore.
